Dev individual learner

Prashant-Android opened the pull request:
- Dispature.IO has been removed from all modules as it is not used in UstadAccountManager.kt.
- Individual screen components have been replaced with ListItem.
- Material Design 3 has been used.
- In App.kt, if(useBottomBar && currentSession?.endpoint?.isLocal != true) has been used.
- All string values are in lowercase.

Prashant-Android committed:
..........

Prashant-Android committed:
Merge branch 'refs/heads/primary' into dev-individual-learner

Prashant-Android committed:
Added 3 button in onboarding screen, created individual login feature

Prashant-Android committed:
Merge remote-tracking branch 'origin/dev-individual-learner' into dev-individual-learner

Prashant-Android committed:
1) Add createLocalAccount() function in UstadAccountManager.kt
2) Create IndividualLearnerScreen and corresponding viewmodel
3) Introduce Dispatcher.IO for common operations
4) Implement 3 Buttons on Onboarding Screen

Prashant-Android committed:
1) Add createLocalAccount() function in UstadAccountManager.kt
2) Create IndividualLearnerScreen and corresponding viewmodel
3) Introduce Dispatcher.IO for common operations
4) Implement 3 Buttons on Onboarding Screen

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/account/UstadAccountManager.kt:
```diff
@@ -505,6 +526,56 @@ class UstadAccountManager(
+        }
+    }
+
+    private fun randomString(length: Int): String {
```
We already have a randomString function - see RandomString.kt . Please use the existing function.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/impl/appstate/AppUiState.kt:
```diff
@@ -24,7 +24,7 @@ data class AppUiState(
     val title: String? = null,
     val navigationVisible: Boolean = true,
-    val hideBottomNavigation: Boolean = false,
+    var hideBottomNavigation: Boolean = false,
```
This should not be changed. AppUiState should be immutable in order to avoid side effects.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/OnBoardingViewModel.kt:
```diff
@@ -53,14 +54,24 @@ class OnBoardingViewModel(
 
+
+    fun onClickIndividual() {
+        navController.navigate(IndividualLearnerViewModel.DEST_NAME, buildMap {
```
Instead of using buildmap when this is empty, use emptyMap() (if this is required).

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/OnBoardingViewModel.kt:
```diff
@@ -53,14 +54,24 @@ class OnBoardingViewModel(
+        })
+    }
+    fun onClickAddNewOrganization() {
+        //Redirect to website (ask mike that is there prebuild util class to open website)
```
This can be done using OpenExternalLinkUseCase . You can use the IDE to find usages of it. You can get it using the dependency injection.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/individual/IndividualLearnerViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    fun onClickRestoreFile() {
+        CoroutineScope(Dispatchers.Main).launch {
```
What is the coroutinescope doing here? 

This needs to take an argument for the file Uri.

The view will need to launch the file browser using the platform itself as is done in other places (see rememberUstadFilePickLauncher. For now, please just change the arguments and remove the (unused) CoroutineScope here.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/individual/IndividualLearnerViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    companion object {
+        const val DEST_NAME = "individualLearner"
```
DEST_NAME value should be capitalized e.g. IndividualLearner

mikedawson commented on core/src/commonMain/resources/MR/base/strings.xml:
```diff
@@ -160,6 +160,18 @@
     <string name="onboarding_subheadline3">Education content, assignments, classes, and attendance all in one place</string>
     <string name="download_continue_stacked_label">Continue download</string>
+
+    <string name="onboarding_individual_text">I\'m an individual learner</string>
```
Please make the string match at least the first few words e.g. name="im_an_individual_learner". 

These are basically new strings. We don't use the prefixes anymore (because that makes it harder to find strings to reuse).

mikedawson commented on core/src/jsMain/kotlin/com/ustadmobile/core/MR.js.kt:
```diff
@@ -0,0 +1,2245 @@
+package com.ustadmobile.core
```
This looks like something created by the IDE in response to a prompt. Please remove it.

This file will be generated for Javascript by the Moko gradle plugin when the Javascript build runs.

This file will break the javascript complation.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/components/UstadHorizontalDivider.kt:
```diff
@@ -0,0 +1,28 @@
+package com.ustadmobile.libuicompose.components
```
This should be removed. Jetpack compose has a [HorizontalDivider](https://developer.android.com/develop/ui/compose/components/divider). We don't need to add our own

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/individual/IndividualLearnerScreen.kt:
```diff
@@ -0,0 +1,112 @@
+fun UstadTitleDescriptionButton(
+    title: String,
+    description: String,
+    titleColor: Color = Color.Black,
```
Avoid setting colors manually as far as possible. We should let the theme do its job.

With this code, if the user is in dark mode, this will go wrong.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/individual/IndividualLearnerScreen.kt:
```diff
@@ -0,0 +1,112 @@
+}
+
+@Composable
+fun UstadTitleDescriptionButton(
```
I don't this function should exist. We can use ListItem as it is.

We should use the default color and sizes provided by the theme unless we have a real need to change them. That makes sure that everything works in dark mode, light mode, looks consistent, etc.

The clickable modifier should be applied to the list item for the click / ripple effect to work as expected.

mikedawson commented on lib-ui-compose/src/desktopMain/kotlin/com/ustadmobile/libuicompose/images/UstadImagePainter.kt:
```diff
@@ -5,6 +5,7 @@ import androidx.compose.ui.graphics.painter.Painter
 import androidx.compose.ui.res.painterResource
 
 val nameMap = mapOf(
+    UstadImage.GOOGLE_ICON to "/img/icn_google.png",
```
As we're not using the Google icon at this stage, let's remove that

Prashant-Android committed:
1) Add createLocalAccount() function in UstadAccountManager.kt
2) Create IndividualLearnerScreen and corresponding viewmodel
3) Introduce Dispatcher.IO for common operations
4) Implement 3 Buttons on Onboarding Screen

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/individual/IndividualLearnerViewModel.kt:
```diff
@@ -0,0 +1,83 @@
+        viewModelScope.launch {
+            accountManager.createLocalAccount()
+            goToNextDestAfterLoginOrGuestSelected(
+                accountManager.createLocalAccount().userSession.usPersonUid,
```
You're calling createLocalAccount three times. We want to create one account, not three.

Prashant-Android replied:
so now its look like this , 
  fun onClickContinueWithoutLogin() {
        loadingState = LoadingUiState.INDETERMINATE

        viewModelScope.launch {
            val localAccount = accountManager.createLocalAccount()
            goToNextDestAfterLoginOrGuestSelected(localAccount.userSession.usPersonUid, localAccount.endpoint)
        }
    }

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/app/App.kt:
```diff
@@ -118,6 +127,16 @@ fun App(
         }
     }
 
+    val isLocalSession = currentSession?.endpoint?.isLocal == true
```
Navigation should be done by the ViewModel (with very few exceptions e.g. if a login expires). This code could have side unforeseen side effects.

The individual user view model should be updated to simply go directly to ContentEntryListViewModel.DEST_NAME_HOME  and this section should be removed.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/onboarding/OnboardingScreen.kt:
This UI should be updated to match the updated design as provided by Akshat e.g. the carousel should be removed. Choices should be in ListItem with the supporting text as per the design.

Prashant-Android committed:
Merge branch 'refs/heads/primary' into dev-individual-learner

Prashant-Android committed:
1) Removed carousel, List Item added.
2) Removed Unused images.
3) Navigate local user from IndividualLearnerViewModel.

Prashant-Android committed:
- Added Zip and Unzip use cases in commonMain and androidMain.
- Implemented FolderSelector utility classes for file and directory selection.
- Integrated zip use cases into DI for commonMain and androidMain.

mikedawson commented on core/src/androidMain/kotlin/com/ustadmobile/core/domain/backup/AndroidUnzipFileUseCase.kt:
```diff
@@ -0,0 +1,68 @@
+import java.io.FileInputStream
+import java.util.zip.ZipInputStream
+
+class AndroidUnzipFileUseCase : UnzipFileUseCase {
```
This should be in commonJvmMain : the zip classes on Android and JVM are the same.

Prashant-Android replied:
How will get this imports 

- import android.content.Context
- import android.net.Uri

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/model/FileToZip.kt:
```diff
@@ -0,0 +1,7 @@
+package com.ustadmobile.core.model
+
+
+data class FileToZip(
```
This data class should be within the UseCase interface - as per coding style doc.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/impl/di/CommonDomainModule.kt:
```diff
@@ -30,6 +35,10 @@ import org.kodein.di.singleton
  */
 fun commonDomainDiModule(endpointScope: EndpointScope) = DI.Module("CommonDomain") {
+
+    bind<ZipFileUseCase>() with singleton { createZipFileUseCase() }
```
The DI binding for a UseCase which has different implementations on different platforms (or no implementation on certain platforms) should be done in platform specific DI - that is also now documented on the coding style guidelines. 

Please use the platform specific DI and remove the expect/actual for creation e.g. createZipFileUseCase etc.

Prashant-Android committed:
-Modifications in rememberUstadFilePickLauncher for folder picking, and updates to other classes accordingly.
-Zip file use cases implemented in commonJvmMain.

Prashant-Android committed:
Added a "share app" option setting screen, changes in IndividualLearnerViewModel while navigating local user to ContentEntryListViewModel added some args to visible import content option in library screen.

Prashant-Android committed:
Change icon to given illustration, fix IndividualLearner library screen content button visibility.

Prashant-Android committed:
Change icon to given illustration, fix IndividualLearner library screen content button visibility.

Prashant-Android committed:
Change icon to given illustration, fix IndividualLearner library screen content button visibility.

Prashant-Android committed:
Change icon to given illustration, fix IndividualLearner library screen content button visibility.

Prashant-Android committed:
Merge branch 'refs/heads/primary' into dev-individual-learner

# Conflicts:
#	app-android/src/main/AndroidManifest.xml
#	app-android/src/main/res/values/strings.xml

Prashant-Android committed:
Merge branch 'refs/heads/primary' into dev-individual-learner

Prashant-Android committed:
Export functionality with JVM Test cases .

Prashant-Android committed:
Export functionality with JVM Test cases .

Prashant-Android committed:
import export code.

Prashant-Android committed:
import export code.

Prashant-Android committed:
Remove offline sharing and import export part.this branch only have individual functionality

mikedawson commented:
Closing as individual / personal mode functionality is now available, and included in dev-onboarding-simplification
